core/state-db/src/pruning.rs

```diff
@@ -193,7 +179,10 @@ impl<BlockHash: Hash, Key: Hash> RefWindow<BlockHash, Key> {
 
 	/// Revert all pending changes
 	pub fn revert_pending(&mut self) {
-		self.pending_records.clear();
+		self.death_rows.truncate(self.death_rows.len() - self.pending_finalizatons);
+		let new_max_block = self.death_rows.len() as u64 + self.pending_number;
+		self.death_index.retain(|_, block| *block < new_max_block);
```

Comment:
If I understand correctly we rollback the applyied tx from earlier `import` call but cannot put back the reinserted removed deathrows from loop under comment `remove all re-inserted keys from death rows`.
Thus your comment in the PR about non pruned values. Should be possible to get a buff of those removed values, is it worth the complexity, I cannot really say (seems hard to exploit as you need to triggered `revert_pending` which is not easy I believe).

Reply:
I'd not bother with it for now.

Reply:
`revert_pending` should never happen under normal conditions.  Transaction failure is pretty much unrecoverable now. This is more of a defensive programming measure.